text/0000-refs.md

```diff
@@ -0,0 +1,209 @@
+</template>
+```
+
+```html
+<template lwc:render-mode="light">
+    <slot lwc:ref="foo"></slot> <!-- Not allowed! -->
+</template>
+```
+
+### `this.refs`
+
+`this.refs` is a plain object, with string keys and DOM node values. If a ref does not exist, then `this.refs["unknown"]` returns `undefined`.
+
+The object is read-only, and attempting to add, modify, or delete its properties from within the component will result in a runtime error.
+
+In the case of shadow DOM components, `this.refs` refers to elements inside of the shadow DOM. Whereas with light DOM components, it refers to elements inside of the light DOM. This makes sense, since in either case, the element must be defined in the template owned by that component in order to be accessible with `this.refs`.
+
+### Multiple templates
+
+In the case of multiple templates, the same rules as for `this.template.querySelector()` or `this.querySelector()` apply. Once the new template is rendered, `this.refs` refers to elements defined in the new template:
+
+```js
+import a from './a.html';
+import b from './b.html';
+
+export default class extends LightningElement {
+  count = 0;
+  
+  render() {
+    return this.count % 2 === 0 ? a : b;
+  }
+  
+  renderedCallback() {
+    console.log(this.refs);
+  }
+  
+  increment () {
+    this.count++;
+  }
+}
+```
+
+```js
+// Logs `this.refs` for a.html
+const cmp = createElement('x-component', { is: Component });
+
+// Logs `this.refs` for b.html
+cmp.increment();
+
+// Logs `this.refs` for a.html
+cmp.increment();
+```
+
+### Duplicate `lwc:ref`s
+
+If duplicate `lwc:ref`s are defined in the same template:
+
+```html
+<template>
+    <div lwc:ref="foo"></div>
+    <div lwc:ref="foo"></div>
+</template>
+```
+
+... then the second one "wins" (in terms of depth-first tree order). The reason for allowing duplicates is explained below.
+
+### Conditional elements
+
+In some cases, it may be useful to have elements that are conditionally defined based on `<template if:true>`. For example:
+
+```html
+<template>
+    <template if:true={darkMode}>
+        <div lwc:ref="foo"></div>      
+    </template>
+    <template if:false={darkMode}>
+        <div lwc:ref="foo"></div>
+    </template>
+</template>
+```
+
+In this case, it's convenient to be able to call `this.refs.foo` and to not care whether the element is inside the first `<template>` or the second.
+
+Because the template compiler cannot know which conditions will cause an element to exist or not, we cannot support this use case while also disallowing duplicates. Therefore, the simplest solution for duplicates is to define whichever element is last (in terms of depth-first tree order) to be the one that "wins."
+
+Here is how other frameworks do it:
+
+- React: [duplicates allowed, last one wins](https://codepen.io/nolanlawson-the-selector/pen/dyVvZwZ)
+- Vue: [duplicates allowed, last one wins](https://codepen.io/nolanlawson-the-selector/pen/jOGBadg)

```

Comment:
For what it's worth, Vue also accepts a function with `:ref` directive to collect references in an array ([v-for array refs](https://v3.vuejs.org/guide/migration/array-refs.html#frontmatter-title) / [example](https://codepen.io/pmdartus/pen/ZEXeZyB)).

Reply:
I think we need some answer to this, yes. There is also the question of what happens with a `ref` inside of a loop. I'll add it to the unanswered questions.